Add Jenkins CI for CCR Performance tests

ankitkala opened the pull request:
Signed-off-by: Ankit Kala <ankikala@amazon.com>

### Description
- Adds the JenkinsFile for triggering CCR Performance tests.
- Add support for creating multi node clusters via `PerfTestCluster`

### Issues Resolved
[1500](https://github.com/opensearch-project/opensearch-build/issues/1500)
 
### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

dblock commented on src/test_workflow/perf_test/perf_test_cluster.py:
```diff
@@ -57,22 +46,31 @@ def start(self):
+            self.endpoint = host
+            self.cluster_endpoint_with_port = "".join([scheme, host, ":", str(self.port())])
+
+    def endpoint_with_port(self):
```
Make this a `@property` so you can use it without `()`.

ankitkala replied:
done.

dblock commented on src/test_workflow/perf_test/perf_test_cluster.py:
```diff
@@ -13,32 +13,20 @@
+    def __init__(self, bundle_manifest, config, stack_name, cluster_config, current_workspace):
         self.manifest = bundle_manifest
-        self.work_dir = os.path.join(current_workspace, "opensearch-cluster", "cdk", "single-node")
+        self.work_dir = os.path.join(current_workspace, "opensearch-cluster", "cdk",
```
You should subclass `PerfTestCluster` into single vs. multi-node. All the methods such as `create_endpoint` and `setup_cdk_params` have different implementations in those cases.

ankitkala replied:
done.

owaiskazi19 commented on src/test_workflow/perf_test/perf_single_node_cluster.py:
```diff
@@ -0,0 +1,55 @@
+from test_workflow.perf_test.perf_test_cluster import PerfTestCluster
+
+
+class PerfSingleNodeCluster(PerfTestCluster):
```
How about if we can create one base class for the PerfCluster and pass the argument as `SingleNode` or `MultiNode` to avoid the duplication?

dblock replied:
The logic is totally different between single-node and multi-node cluster, that's why there are 2 child classes here.

dblock commented on src/test_workflow/perf_test/perf_single_node_cluster.py:
```diff
@@ -0,0 +1,55 @@
+        public_ip = cdk_output[self.stack_name].get("PublicIp", None)
+        self.is_endpoint_public = public_ip is not None
+        host = public_ip if public_ip is not None else private_ip
+        if host is not None:
```
Isn't this an error in the case where `host` is `None`? What should happen here? Default to localhost?

ankitkala replied:
Yep. Ideally this shouldn't happen. I'll remove the 'None' check.

dblock commented on src/test_workflow/perf_test/perf_test_cluster.py:
```diff
@@ -87,23 +72,13 @@ def dependencies(self):
-        url = "".join([self.endpoint(), "/_cluster/health"])
+        # Should be invoked only if the endpoint is public.
+        assert self.is_endpoint_public, "wait_for_processing should be invoked only when cluster is public"
+        print("Waiting for domain to be up")
```
Use `logging.info`

dblock commented on src/test_workflow/perf_test/perf_test_cluster_config.py:
```diff
@@ -0,0 +1,20 @@
+        self.ingest_nodes = ingest_nodes
+        self.client_nodes = client_nodes
+
+    def is_single_node_cluster(self):
```
Make it a `@property` and maybe just call it `single_node_cluster`

ankitkala replied:
done

dblock commented on src/test_workflow/perf_test/perf_test_cluster_config.py:
```diff
@@ -0,0 +1,20 @@
+        self.client_nodes = client_nodes
+
+    def is_single_node_cluster(self):
+        return True if (self.data_nodes == 1 and self.master_nodes == 0
```
return `self.data_nodes == ...`, no need to do `if/else`, it's already a bool.

dblock reviewed (approved):
This is a lot better. 

I have some smaller asks that you might want to consider before merging.

kotwanikunal commented on jenkins/cross-cluster-replication/perf-test.jenkinsfile:
```diff
@@ -0,0 +1,138 @@
+        timeout(time: 10, unit: 'HOURS')
+    }
+    environment {
+        AGENT_LABEL = 'Jenkins-Agent-al2-x64-c54xlarge-Docker-Host'
```
Please use the agent label - `Jenkins-Agent-al2-x64-m5xlarge-Perf-Test`

ankitkala replied:
ack.

ankitkala commented:
> This is a lot better.
> 
> I have some smaller asks that you might want to consider before merging.
> 
> Also fix CI.

Addressed all the comments and fixed the broken tests.

gaiksaya commented on jenkins/cross-cluster-replication/perf-test.jenkinsfile:
```diff
@@ -0,0 +1,139 @@
+lib = library(identifier: "jenkins@20211118", retriever: legacySCM(scm))
```
Do we  plan to add regression tests for this jenkins file? Similar to https://github.com/opensearch-project/opensearch-build/tree/main/tests/jenkins/jenkinsjob-regression-files
More details on how to add: https://github.com/opensearch-project/opensearch-build/blob/main/DEVELOPER_GUIDE.md#regression-tests

ankitkala replied:
Added the test for the jenkins file

gaiksaya commented on src/test_workflow/perf_test/perf_test_cluster.py:
```diff
@@ -57,25 +45,22 @@ def start(self):
 
     def terminate(self):
-        os.chdir(os.path.join(self.current_workspace, self.work_dir))
+        os.chdir(self.work_dir)
```
terminate can also be converted to `@property` right?

dblock replied:
It shouldn't. A property is a field/noun (e.g. `object.color` or `object.shape`), invoked without `()` vs. `terminate` is an action/verb, invoked with `()` (e.g. `object.run()`, `object.stop()`).

gaiksaya reviewed (commented):
Looks like there are some failing tests for windows.

dblock commented on src/test_workflow/perf_test/perf_multi_node_cluster.py:
```diff
@@ -0,0 +1,54 @@
+            cluster.start()
+            yield cluster
+        finally:
+            cluster.terminate()
```
This method should be in the base `PerfTestCluster` class.

ankitkala replied:
My bad. Had to move this to subclass during dev testing and forgot to move it back.

dblock commented on src/test_workflow/perf_test/perf_single_node_cluster.py:
```diff
@@ -0,0 +1,53 @@
+            cluster.start()
+            yield cluster
+        finally:
+            cluster.terminate()
```
Duplicate, move to base class.

ankitkala replied:
Done

dblock commented on src/test_workflow/perf_test/perf_multi_node_cluster.py:
```diff
@@ -0,0 +1,54 @@
+        scheme = "https://" if self.cluster_config.security else "http://"
+        host = cdk_output[self.stack_name]["LoadBalancerEndpoint"]
+        self.is_endpoint_public = True
+        if host is not None:
```
I still question what happens here if `host == None`. Is that valid? We end up without values in `endpoint`, is that expected?

ankitkala replied:
Yes, we'd be without value in endpoint. Not sure if there is any graceful way to handle that. 

I'm thinking we can assert on the host not being `None` instead. Thoughts?

dblock replied:
Raise a `ValueError`

dblock reviewed (changes requested):
CI is still failing + some more small stuff

dblock commented on src/test_workflow/perf_test/perf_single_node_cluster.py:
```diff
@@ -0,0 +1,44 @@
+        public_ip = cdk_output[self.stack_name].get("PublicIp", None)
+        self.is_endpoint_public = public_ip is not None
+        host = public_ip if public_ip is not None else private_ip
+        assert host is not None, "Unable to fetch the cluster endpoint from cdk output"
```
I think this error should happen at runtime, there's a legit cdk output that might be missing this data, so I think it's not an assert, but a normal exception. And would need a test that it's raised.

ankitkala replied:
Sure. Changed to RuntimeError and added the tests for it.
